_source/_posts/2022-03-29-github-actions-graalvm.adoc

```diff
@@ -0,0 +1,490 @@
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v1
+        with:
+          node-version: 16.14.0
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'temurin'
+          java-version: 11
+      - name: Install node.js packages
+        run: npm install
+      - name: Run backend test
+        run: |
+          chmod +x mvnw
+          npm run ci:backend:test
+      - name: Run frontend test
+        run: npm run ci:frontend:test
+      - name: Package application
+        run: npm run java:jar:prod
+      - name: 'E2E: Package'
+        run: npm run ci:e2e:package
+      - name: 'E2E: Prepare'
+        run: npm run ci:e2e:prepare
+      - name: 'E2E: Run'
+        run: npm run ci:e2e:run
+        env:
+          CYPRESS_ENABLE_RECORD: false
+          CYPRESS_PROJECT_ID: ${{ secrets.CYPRESS_PROJECT_ID }}
+          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
+      - name: 'E2E: Teardown'
+        run: npm run ci:e2e:teardown
+----
+{% endraw %}
+
+To test this out on your new repository, you'll need to create a branch and pull request (PR) with your changes.
+
+[source,shell]
+----
+git checkout -b actions
+git add .
+git commit -m "Add GitHub Actions"
+git push ci-cd actions
+----
+
+If you watch the tests run from your PR, you'll be pretty pleased until it hits the **E2E: Package** phase. It'll fail with the following error:
+
+----
+Error:  Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.9.10:build
+  (build-native) on project flickr-2: Execution build-native of goal
+  org.graalvm.buildtools:native-maven-plugin:0.9.10:build failed:
+  GraalVM native-image is missing from your system.
+Error:   Make sure that GRAALVM_HOME environment variable is present.
+----
+
+== The environmental impact of GraalVM builds
+
+This brings us to a fork in the road. Should you use the https://github.com/graalvm/setup-graalvm/issues/6[setup-graalvm action] to configure GraalVM and your Java SDK?
+
+If you do, every time you create a PR, and commit to it, it will run a native build. These take a 3-4 minutes for me locally, but with GitHub Actions, it takes 30+ minutes! Also, it only builds on Linux, so you won't know if there are issues with macOS or Windows.
+
+To me, this seems as bad for the environment as cryptocurrency. If you're using a private repo, it'll also make you wish you bought crypto a couple of years ago. You only get 2000 free minutes of GitHub Actions for private repos. Any minutes after that, you get charged for.
+
+_Yes, I know the cryptocurrency topic is controversial. I do like to poke fun at it though. Native builds on every commit and mining bitcoin seem similar to me._
+
+== Continuously test JVM builds with GitHub Actions
+
+There's a better way, in my opinion. Revert the changes made by the JHipster Native blueprint to the tasks that are run by the `npm run` commands. That is, change `ci:e2e:package` from `npm run native-package` to the following:
```

Comment:
Due to required quirks that may be randomly required by native build, you could recommend a native executable daily build.
This could avoid a big bisect to find the problem.

Reply:
And we may keep ci scripts untouched and change the blueprint workflow to use the custom scripts.

Reply:
> Due to required quirks that may be randomly required by native build, you could recommend a native executable daily build.

@mshima There is a sentence in the post:

> If you want to test things more often, you could do it with a [nightly build](https://kyuubang.github.io/devops/2022/01/22/nightly-build-with-github-actions.html).

As far as renaming `github-actions.yml` to `main.yml`, that should be in the next release (v7.7.1). I'll remove this part and test this tutorial again after it's released.

For the other things (e.g. keeping ci scripts untouched), I can pause the publishing of this post until after the next release of JHipster Native, or I can update this post (and the one it depends on) after it's released.

Reply:
> If you want to test things more often, you could do it with a [nightly build](https://kyuubang.github.io/devops/2022/01/22/nightly-build-with-github-actions.html).

I saw, but it's in a different context, maybe move it here then.